packages/trace-viewer/src/ui/uiModeView.tsx

```diff
@@ -83,62 +79,129 @@ export const UIModeView: React.FC<{}> = ({
+      },
+      pathSeparator,
+    });
+
+    const updateList = async () => {
+      setIsLoading(true);
+      const result = await testServerConnection.listTests({});
+      teleSuiteUpdater.processListReport(result.report);
+      setIsLoading(false);
+    };
+
+    setTestModel(undefined);
     setIsLoading(true);
-    reloadTests();
-  }, [reloadTests]);
+    setWatchedTreeIds({ value: new Set() });
+    (async () => {
+      const status = await testServerConnection.runGlobalSetup();
+      if (status !== 'passed')
+        return;
+      const result = await testServerConnection.listTests({});
+      teleSuiteUpdater.processListReport(result.report);
+
+      testServerConnection.onListChanged(updateList);
+      testServerConnection.onReport(params => {
+        teleSuiteUpdater.processTestReport(params);
+      });
+      setIsLoading(false);
+
+      const { hasBrowsers } = await testServerConnection.checkBrowsers();
+      setHasBrowsers(hasBrowsers);
+    })();
+  }, [testServerConnection]);
 
-  updateRootSuite = React.useCallback((config: reporterTypes.FullConfig, rootSuite: reporterTypes.Suite, loadErrors: reporterTypes.TestError[], newProgress: Progress | undefined) => {
+  // Update project filter default values.
+  React.useEffect(() => {
+    if (!testModel)
+      return;
+
+    const { config, rootSuite } = testModel;
     const selectedProjects = config.configFile ? settings.getObject<string[] | undefined>(config.configFile + ':projects', undefined) : undefined;
-    for (const projectName of projectFilters.keys()) {
+    const newFilter = new Map(projectFilters);
+    for (const projectName of newFilter.keys()) {
       if (!rootSuite.suites.find(s => s.title === projectName))
-        projectFilters.delete(projectName);
+        newFilter.delete(projectName);
     }
     for (const projectSuite of rootSuite.suites) {
-      if (!projectFilters.has(projectSuite.title))
-        projectFilters.set(projectSuite.title, !!selectedProjects?.includes(projectSuite.title));
+      if (!newFilter.has(projectSuite.title))
+        newFilter.set(projectSuite.title, !!selectedProjects?.includes(projectSuite.title));
     }
-    if (!selectedProjects && projectFilters.size && ![...projectFilters.values()].includes(true))
-      projectFilters.set(projectFilters.entries().next().value[0], true);
+    if (!selectedProjects && newFilter.size && ![...newFilter.values()].includes(true))
+      newFilter.set(newFilter.entries().next().value[0], true);
+    if (projectFilters.size !== newFilter.size || [...projectFilters].some(([k, v]) => newFilter.get(k) !== v))
+      setProjectFilters(newFilter);
+  }, [projectFilters, testModel]);
 
-    setTestModel({ config, rootSuite, loadErrors });
-    setProjectFilters(new Map(projectFilters));
-    if (runningState && newProgress)
-      setProgress(newProgress);
-    else if (!newProgress)
+  // Update progress.
+  React.useEffect(() => {
+    if (runningState && testModel?.progress)
```

Comment:
I don't really understand this logic, but I guess it's the same as before...

Reply:
No change here.